### PR TITLE
test: Add unit tests for directory policy watcher

### PR DESCRIPTION
**Pull request by tamilmani1989**
Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

related: #33160

### CODE REVIEW

**Comment by tamilmani1989** [comment]
cc: @gandro

File: pkg/policy/directory/watcher_test.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	require.Equal(t, 0, len(p.fileNameToCnpCache))
+
+	// Delete non existent entry and validate if appropriate error returned
+	p.addToPolicyEngine(cnp, "test2.yaml")
+	err = p.deleteFromPolicyEngine("test.yaml")
+	require.ErrorContains(t, err, "fileNameToCnp map entry doesn't exist",
+		"Expected error `fileNameToCnp map entry doesn't exist` but got:%v", err)
+	require.Equal(t, 1, len(p.fileNameToCnpCache))
+	err = p.deleteFromPolicyEngine("test2.yaml")
+	require.NoError(t, err, "Expect no error while delete policy but got error:%v", err)
```
Comment (doniacld) [inline pkg/policy/directory/watcher_test.go:138]:
```suggestion
	require.NoError(t, err, "Expect no error while deleting policy but got error:%v", err)
```

**Review by doniacld — APPROVED** [review]
One nit comment otherwise LGTM!

**Review by gandro — COMMENTED** [review]
Thanks for adding the unit test. Have you considered actually using a temporary filesystem for testing? That way, we really test the watcher functionality, which at the moment is not tested otherwise. 

For some inspiration, see the certloader package which observes a folder for certificate files and loads them into the certificate store: https://github.com/cilium/cilium/blob/5ec17e8200baf816cffccdc832ce5ac05dc129c2/pkg/crypto/certloader/watcher_test.go#L153-L215

**Comment by tamilmani1989** [comment]
> Thanks for adding the unit test. Have you considered actually using a temporary filesystem for testing? That way, we really test the watcher functionality, which at the moment is not tested otherwise.
> 
> For some inspiration, see the certloader package which observes a folder for certificate files and loads them into the certificate store:
> 
> https://github.com/cilium/cilium/blob/5ec17e8200baf816cffccdc832ce5ac05dc129c2/pkg/crypto/certloader/watcher_test.go#L153-L215

for unit test, I thought better to abstract platform/system calls and just test the functionality for various inputs.. thats why redesigned the code a bit to separate filesystem calls and not testing system calls

**Comment by gandro** [comment]
> for unit test, I thought better to abstract platform/system calls and just test the functionality for various inputs.. thats why redesigned the code a bit to separate filesystem calls and not testing system calls

While I agree in principle, I feel the most complex part of this piece of code is the file watcher logic and I strongly believe we would benefit from some tests that ensure that file system operations do indeed have the desired effect on the loaded policies. In particular since other packages in Cilium using `fsnotify` have had subtle bugs that are very hard to reproduce in our end-to-end testing framework, but trivial to trigger by setting up a temporary filesystem in a `go test` run,

We don't have to call them "unit tests", in fact we often use `_test.go` files in Cilium for integration tests too (see the `integration-tests` Makefile target).

In any case, I will not block the PR on; this point, since this is already a clear improvement as is.

File: pkg/policy/directory/watcher.go [hunk 2]
```diff
@@ -124,7 +122,7 @@ func (p *policyWatcher) deleteFromPolicyEngine(cnpFilePath string) error {
 	cnp := p.fileNameToCnpCache[fileName]
 	if cnp == nil {
 		p.log.WithField("file", fileName).Error("BUG: Policy deletion request for file which was never added")
```
Comment (gandro) [inline pkg/policy/directory/watcher.go]: we're now logging this error twice (once here and once at the call-site). If we return the error here, we probably can get rid of this log statement

File: pkg/policy/directory/watcher.go [hunk 3]
```diff
@@ -34,12 +35,7 @@ type policyWatcher struct {
 	fileNameToCnpCache map[string]*cilium_v2.CiliumNetworkPolicy
 }
 
-func (p *policyWatcher) translateToCNPObject(file string) (*cilium_v2.CiliumNetworkPolicy, error) {
-	data, err := os.ReadFile(file)
-	if err != nil {
-		return nil, err
-	}
-
+func (p *policyWatcher) translateToCNPObject(data []byte) (*cilium_v2.CiliumNetworkPolicy, error) {
```
Comment (gandro) [inline pkg/policy/directory/watcher.go:38]: Have you considered mocking the filesystem with https://pkg.go.dev/io/fs#FS

This way, you could pass in a filepath and a `fs.FS` reference here. You can obtain a real FS via https://pkg.go.dev/os#DirFS and for testing use a  https://pkg.go.dev/testing/fstest#MapFS ?

**Review by gandro — COMMENTED** [review]
Two minor nits

**Comment by tamilmani1989** [comment]
> > for unit test, I thought better to abstract platform/system calls and just test the functionality for various inputs.. thats why redesigned the code a bit to separate filesystem calls and not testing system calls
> 
> While I agree in principle, I feel the most complex part of this piece of code is the file watcher logic and I strongly believe we would benefit from some tests that ensure that file system operations do indeed have the desired effect on the loaded policies. In particular since other packages in Cilium using `fsnotify` have had subtle bugs that are very hard to reproduce in our end-to-end testing framework, but trivial to trigger by setting up a temporary filesystem in a `go test` run,
> 
> We don't have to call them "unit tests", in fact we often use `_test.go` files in Cilium for integration tests too (see the `integration-tests` Makefile target).
> 
> In any case, I will not block the PR on; this point, since this is already a clear improvement as is.

make sense. I will create a separate PR for this

File: pkg/policy/directory/watcher.go (hunk 3 above)
Comment (tamilmani1989) [reply pkg/policy/directory/watcher.go:38]: will take a look

**Review by gandro — APPROVED** [review]

**Comment by gandro** [comment]
Thank you!

**Comment by gandro** [comment]
Rebased on main

**Pull request merged**